### ISSUE TITLE
Incorrect quote string matching

### ISSUE THREAD

**Issue by scagood** [spencermountain/compromise#458]
## Overview

The quotation string matching system matches incorrectly for adjacent strings.

## Example

```js
const nlp = require('compromise');

// Try splitting the string into quoted strings.
console.log(
    nlp('My "String" "with many" adjacent "nested" \'quotes\'')
    .quotations()
    .out('array')
)
```
#### Output
```json
[
    "string with many",
    "nested quotes"
]
```
#### Expected
```json
[
    "string",
    "with many",
    "nested",
    "quotes"
]
```
### Workaround
So, I solved this for my specific circumstance like so:
```js
function quotes(string) {
    // Match a single word surrounded by (")
    return string.match(
        '/^"[^"\\\\]*(?:\\\\.[^"\\\\]*)*"$/'
    )
    
    // Match a single word surrounded by (')
    .concat(string.match(
        '/^\'[^\'\\\\]*(?:\\\\.[^\'\\\\]*)*\'$/'
    ))
    
    
    // Match a two or more word surrounded by (")
    .concat(string.match(
        '/^"[^"\\\\]*(?:\\\\.[^"\\\\]*)*$/ ' +
        '/[^"\\\\]*(?:\\\\.[^"\\\\]*)*/+? ' +
        '/^[^"\\\\]*(?:\\\\.[^"\\\\]*)*"$/'
    ))
    
    // Match a two or more word surrounded by (')
    .concat(string.match(
        '/^\'[^\'\\\\]*(?:\\\\.[^\'\\\\]*)*$/ ' +
        '/[^\'\\\\]*(?:\\\\.[^\'\\\\]*)*/+? ' +
        '/^[^\'\\\\]*(?:\\\\.[^\'\\\\]*)*\'$/'
    ));
}
```
Using the RegExp by [Jeffrey Friedl](http://rads.stackoverflow.com/amzn/click/0596528124) provided by [ridgerunner](https://stackoverflow.com/u/433790) on [stackoverflow](https://stackoverflow.com/a/5696141/3533202)

**Edit 1**
I will note my solution is not perfect as it outputs:
```json
[
    "string",
    "nested",
    "quotes",
    "with many"
]
```
**End:  Edit 1**

> A side note:
> The regex on the [match syntax wiki page](https://github.com/spencermountain/compromise/wiki/Match-Syntax) should be: `/rain(ing|ed)/` meaning match `rain` and (`ing` or `ed`).
> The current `/rain[ing|ed]/` means match `rain` and (`i`, `n`, `g`, `|` `e` or `d`).

**Comment by spencermountain** [comment]
ah, thanks Sca. This is a great bug. and your workaround is very clever. 
i think i've fixed this in `11.5.2` - at least your example works. Can you take a look?
There's all sorts of issues with, unicode quoations, um and the trailing possessive apostrophe inside a quote, which might trip it still.
cheers

**Comment by scagood** [comment]
That seems to have worked for my previous method (I've since changed the way I was processing the instruction list).

I will note this may trip up with nested quotations, however, I'm not sure who would use nested quotes. Here is the sort of monstrosity I mean when I say 'nested' quotes: `"'She ate it!?', he said" she said.`. Again, I'm not sure who would do that, or how it would be processed. (v 11.5.2 produces the inner quotes.)

As for possessive trailing apostrophes, this seems to be a touch off in 'odd' situations, or example:
`"I told you, That's James'", someone shrieked from around the corner.`  loses the trailing `'`, but, `'I told you, That's James'', someone shrieked from around the corner.` keeps it.

I have not tried with Unicode characters, I assumed #43 had that covered?

**Comment by spencermountain** [comment]
yeah, there's one pass in [./tagger/steps](https://github.com/spencermountain/compromise/blob/master/src/tagger/steps/18-quotation_step.js) that tags things as quotations, then another step [here](https://github.com/spencermountain/compromise/blob/master/src/text/subsets.js#L44) that splits them into arrays.

happy as always, to merge a pr, if you can make this logic smarter

**Comment by scagood** [comment]
So, I've been thinking, and nested quotes seem out of scope for now. 
Sorting the possessive nouns and quotations seems plausible.

---------
Here are my proposed changes:
#### 1. I believe the '18-quotation_step' would need to happen before the '15-possessive_step'. 
I'll start with an example:
`Here, 'quotation marks' are used to denote both Possessive and Quotation tags.`
Here detecting the whether `marks'` is meant to be possessive is very hard without knowing about quotations.
Here is the current tags for 'marks'
```json
{
  "text": "marks'",
  "normal": "marks",
  "tags": [
    "Possessive",
    "Noun",
    "MaleName",
    "FirstName",
    "Person",
    "Singular",
    "Quotation"
  ]
}
```

Here is the same element tags in the same sentence, just without quotes:
```json
{
  "text": "marks",
  "normal": "marks",
  "tags": [
    "Noun",
    "Plural"
  ]
}
```

#### 2. The opening quote should dictate what type of closing quote to search for.
In the following example, the opening double quote closes on the single quote:
`Here, "quotation marks' are used to denote both "Possessive" and 'Quotation' tags.`
Using `.quotations()` on this would output something like this:
```json
[
  "quotation marks",
  "possessive",
  "quotation"
]
```
That is more obvious in a sentence like so: `Here, "that's marks' cat" will quote "that's marks".`.

Doing that means alternate forms of quotations can be matched correctly (See: http://snowball.tartarus.org/texts/apostrophe.html).

---
Do you have anything else that you may want to add before I see if I can implement this?
Is there a reason that I missed for tagging possessive nouns before quotations?

**Comment by spencermountain** [comment]
YEESSS!

oh my gosh, that apostrophe website makes me sad. I forgot about a lot of those cases they mention.

i'm on board!
lemme know if i can do anything to help. I'm around on [the slack group](https://slackin-kyzvclgjlg.now.sh/) all-day. Happy to work together on this, if you want
cheers

**Comment by scagood** [comment]
This is quite a read, sorry in advance :')
I've code folded a lot of the extra/utility information (Prefixed with a &#11208; to show where they are.)

---

Right, here is my first attempt at tagging the quotation.
If you run this:
```js
nlp(`My "String" "with many" adjacent "nested" 'quotes'`).debug();
nlp(`My "String 'with many' adjacent" "nested" 'quotes'`).debug();
```

<details>
  <summary>Output:</summary>
  <p>

```
====
   --
   'My'                       - TitleCase, Possessive, Noun
   '"String"'                 - OpenQuotation, CloseQuotation, Quotation, StraightDoubleQuotes, Noun
   '"with'                    - Preposition, OpenQuotation, Quotation, StraightDoubleQuotes
   'many"'                    - Adjective, CloseQuotation, Quotation, StraightDoubleQuotes
   'adjacent'                 - Adjective
   '"nested"'                 - Noun, OpenQuotation, CloseQuotation, Quotation, StraightDoubleQuotes, Singular
   ''quotes''                 - Noun, OpenQuotation, CloseQuotation, Quotation, StraightSingleQuotes, Plural
====
   --
   'My'                       - TitleCase, Possessive, Noun
   '"String'                  - OpenQuotation, Quotation, StraightDoubleQuotes, Noun
   ''with'                    - Preposition, Quotation, StraightDoubleQuotes
   'many''                    - Adjective, Quotation, StraightDoubleQuotes
   'adjacent"'                - Adjective, CloseQuotation, Quotation, StraightDoubleQuotes
   '"nested"'                 - Noun, OpenQuotation, CloseQuotation, Quotation, StraightDoubleQuotes, Singular
   ''quotes''                 - Noun, OpenQuotation, CloseQuotation, Quotation, StraightSingleQuotes, Plural
```
  </p>
</details>

However, regarding the conversation we had on slack, and looking at the mess that is the tags, I don't think this is the way forward?

<details>
  <summary>This is the modified 'quotation_step' </summary>
  <p>

```js
'use strict';

const quotemarks = {
    '\u0022': {close: '\u0022', tag: 'StraightDoubleQuotes'},
    '\uFF02': {close: '\uFF02', tag: 'StraightDoubleQuotesWide'},
    '\u0027': {close: '\u0027', tag: 'StraightSingleQuotes'},

    '\u201C': {close: '\u201D', tag: 'CommaDoubleQuotes'},
    '\u2018': {close: '\u2019', tag: 'CommaSingleQuotes'},

    '\u201F': {close: '\u201D', tag: 'CurlyDoubleQuotesReversed'},
    '\u201B': {close: '\u2019', tag: 'CurlySingleQuotesReversed'},

    '\u201E': {close: '\u201D', tag: 'LowCurlyDoubleQuotes'},
    '\u2E42': {close: '\u201D', tag: 'LowCurlyDoubleQuotesReversed'},

    '\u201A': {close: '\u2019', tag: 'LowCurlySingleQuotes'},

    '\u00AB': {close: '\u00BB', tag: 'AngleDoubleQuotes'},
    '\u2039': {close: '\u203A', tag: 'AngleSingleQuotes'},

    // Prime 'non quotation'
    '\u2035': {close: '\u2032', tag: 'PrimeSingleQuotes'},
    '\u2036': {close: '\u2033', tag: 'PrimeDoubleQuotes'},
    '\u2037': {close: '\u2034', tag: 'PrimeTripleQuotes'},

    // Prime 'quotation' variation
    '\u301D': {close: '\u301E', tag: 'PrimeDoubleQuotes'},
    '\u0060': {close: '\u00B4', tag: 'PrimeSingleQuotes'},

    '\u301F': {close: '\u301E', tag: 'LowPrimeDoubleQuotesReversed'}
};

//tag a inline quotation as such
const quotation_step = ts => {
  for (let i = 0; i < ts.terms.length; i++) {
    let t = ts.terms[i];
    if (typeof quotemarks[t.text[0]] === 'object') {
      const quoteInfo = quotemarks[t.text[0][0]];
      const endQuote = new RegExp(quoteInfo.close + '[;:,.]?$');
      
      t.tag('OpenQuotation', 'quotation_step');

      //look for the ending
      for (let o = 0; o < ts.terms.length; o++) {
        //max-length- don't go-on forever
        if (!ts.terms[i + o] || o > 28) {
          break;
        }

        if (endQuote.test(ts.terms[i + o].text) === true) {
          ts.terms[i + o].tag('CloseQuotation', 'quotation_step');
          ts.slice(i, i + o + 1).tag([
              'Quotation',
              quoteInfo.tag
          ], 'quotation_step');
          i += o;
          break;
        }
      }
    }
  }
  return ts;
};
module.exports = quotation_step;
```
  </p>
</details>

---

So, moving on to the second thing we disused, not tagging quotations at all, while keeping 'tabs' on quotations in the possessive step.

So, I'm just going to append the following to the quotation step temporarily.
```js
module.exports = terms => terms;
```
Meaning the initial nlp example
```js
nlp(`My "String" "with many" adjacent "nested" 'quotes'`).debug();
nlp(`My "String 'with many' adjacent" "nested" 'quotes'`).debug();
```

<details>
  <summary>Would now output:</summary>
  <p>

```
====
   --
   'My'                       - TitleCase, Possessive, Noun
   '"String"'                 - Noun
   '"with'                    - Preposition
   'many"'                    - Adjective
   'adjacent'                 - Adjective
   '"nested"'                 - Noun, Singular
   ''quotes''                 - Noun, Plural
====
   --
   'My'                       - TitleCase, Possessive, Noun
   '"String'                  - Noun
   ''with'                    - Preposition
   'many''                    - Adjective
   'adjacent"'                - Adjective
   '"nested"'                 - Noun, Singular
   ''quotes''                 - Noun, Plural
```
  </p>
</details>

As we previously discussed this would be okay, if it wasn't for strings like this:
```js
nlp(`My String with 'quotation things' and adjacent "nested" 'quotes'`).debug();
```
<details>
  <summary>This outputs the following:</summary>
  <p>

```
   'My'                       - TitleCase, Possessive, Noun
   'String'                   - TitleCase, Noun, Singular
   'with'                     - Preposition
   ''quotation'                - Singular, Noun
   'things''                  - Possessive, Noun, Plural
   'and'                      - Conjunction
   'adjacent'                 - Adjective
   '"nested"'                 - Noun, Singular
   ''quotes''                 - Noun, Plural
```
  </p>
</details>
As the `things'` is seen as possessive, we need to keep 'tabs' on quotation marks in the possessive steps. 

I believe that it's a safe assumption to assume only single quotation marks can be used in the possessive `s'`, meaning we can now use a reduced quotation object like so to detect only the quotations we need to worry about:
```js
const singleQuotes = {
    '\u0027': '\u0027', // Straight Single Quotes
    '\u2018': '\u2019', // Comma Single Quotes
    '\u201B': '\u2019', // Curly Single Quotes Reversed
    '\u201A': '\u2019', // Low Curly Single Quotes
    '\u2039': '\u203A', // Angle Single Quotes
    '\u2035': '\u2032', // Prime Single Quotes
    '\u0060': '\u00B4'  // Prime Single Quotes
};
```

So, I set out with a set of strings I wanted to tag `Possession` on correctly. Those being:
```js
nlp(`I think that is Joe's, yes, it's his.`).debug();
nlp(`I think that is Carlos', yes, it's his.`).debug();
nlp(`They're Mark's 'question marks'`).debug();
nlp(`They're Mark's 'hats'`).debug();

// I will note:
// `They're Mark's 'question marks'` is not tagged quite correctly, and hence wont work in this example.
// The issue here is how `question` is tagged:
// It is tagged: `Infinitive, Verb, ...`.
// It is really: `Noun, singular, ...`
// But that is another problem for another day.
```

Unfortunately I almost immediately encountered a problem, that being the `possessive_step` was not the only place tagging things as possessive. (I couldn't find the other place).

<details>
  <summary>This essentially meant even with both the possessive step and the quotation step commented out the output looked like this:</summary>
  <p>

```
====
   --
   'I'                        - Pronoun, Noun, Singular
   'think'                    - Infinitive, PresentTense, Verb, VerbPhrase
   'that'                     - Determiner
   'is'                       - Copula, Verb, VerbPhrase
   'Joe's,'   [Joe]           - TitleCase, MaleName, FirstName, Person, Singular, Noun, Contraction, Comma, ClauseEnd
   '-'        [is]            - Contraction, Copula, Verb, VerbPhrase
   'yes,'                     - Comma, ClauseEnd, Verb, VerbPhrase
   'it's'     [it]            - Pronoun, Noun, Contraction
   '-'        [is]            - Contraction, Copula, Verb, VerbPhrase
   'his.'                     - Possessive, Noun
====
   --
   'I'                        - Pronoun, Noun, Singular
   'think'                    - Infinitive, PresentTense, Verb, VerbPhrase
   'that'                     - Determiner
   'is'                       - Copula, Verb, VerbPhrase
   'Carlos','                 - TitleCase, Possessive, Noun, MaleName, FirstName, Person, Singular
   'yes,'                     - Adverb, Comma, ClauseEnd, VerbPhrase
   'it's'     [it]            - Contraction, Verb, VerbPhrase
   '-'        [is]            - Contraction, Copula, Verb, VerbPhrase
   'his.'                     - Possessive, Noun
====
   --
   'They're'  [they]          - TitleCase, Pronoun, Noun, Contraction
   '-'        [are]           - Contraction, Copula, Verb, VerbPhrase
   'Mark's'   [Mark]          - TitleCase, MaleName, FirstName, Person, Singular, Noun, Contraction
   '-'        [has]           - Contraction
   ''question'                - Infinitive, PresentTense, Verb, VerbPhrase
   'marks''                   - Possessive, Noun, MaleName, FirstName, Person, Singular
====
   --
   'They're'  [they]          - TitleCase, Pronoun, Noun, Contraction
   '-'        [are]           - Contraction, Copula, Verb, VerbPhrase
   'Mark's'                   - TitleCase, MaleName, FirstName, Person, Singular, Noun, Possessive
   ''hats''                   - Noun, Plural
```
</p>
</details>



<details>
  <summary>The first things I did was remove the previously set possessive tag, leaving an output like so:</summary>
  <p>

```
====
   --
   'I'                        - Pronoun, Noun, Singular
   'think'                    - Infinitive, PresentTense, Verb, VerbPhrase
   'that'                     - Determiner
   'is'                       - Copula, Verb, VerbPhrase
   'Joe's,'   [Joe]           - TitleCase, MaleName, FirstName, Person, Singular, Noun, Contraction, Comma, ClauseEnd
   '-'        [is]            - Contraction, Copula, Verb, VerbPhrase
   'yes,'                     - Comma, ClauseEnd, Verb, VerbPhrase
   'it's'     [it]            - Pronoun, Noun, Contraction
   '-'        [is]            - Contraction, Copula, Verb, VerbPhrase
   'his.'                     - Noun, Singular
====
   --
   'I'                        - Pronoun, Noun, Singular
   'think'                    - Infinitive, PresentTense, Verb, VerbPhrase
   'that'                     - Determiner
   'is'                       - Copula, Verb, VerbPhrase
   'Carlos','                 - TitleCase, Noun, MaleName, FirstName, Person, Singular
   'yes,'                     - Adverb, Comma, ClauseEnd, VerbPhrase
   'it's'     [it]            - Contraction, Verb, VerbPhrase
   '-'        [is]            - Contraction, Copula, Verb, VerbPhrase
   'his.'                     - Noun, Singular
====
   --
   'They're'  [they]          - TitleCase, Pronoun, Noun, Contraction
   '-'        [are]           - Contraction, Copula, Verb, VerbPhrase
   'Mark's'   [Mark]          - TitleCase, MaleName, FirstName, Person, Singular, Noun, Contraction
   '-'        [has]           - Contraction
   ''question'                - Infinitive, PresentTense, Verb, VerbPhrase
   'marks''                   - Noun, MaleName, FirstName, Person, Singular
====
   --
   'They're'  [they]          - TitleCase, Pronoun, Noun, Contraction
   '-'        [are]           - Contraction, Copula, Verb, VerbPhrase
   'Mark's'                   - TitleCase, MaleName, FirstName, Person, Singular, Noun, ProperNoun
   ''hats''                   - Noun, LastName, Person, Singular, ProperNoun
```
</p>
</details>

This has the unfortunate side effect of un-tagging actual possessive determiners and adjectives, for example 'my' and 'his', but this is a simple thing to solve, once the punctuation rule is found.

So, now I can implement the following logic to counter quotes, and other forms of punctuation. I will note, this will not fix adjacent nested single quotes. But it's a very unlikely thing to see adjacent to a possessive. the term would have to look something like this:
```
Some, ‘Utterly, ‹Bizzare, nested quote, `that happens to be Carlos'´›’
```
But, I've never seen this type of... monstrosity... so, will ignore it... for now 😈.

First, the predefined variable for this step will now look similar to this:
```js
// Our singleQuotes object.
const singleQuotes = {
    '\u0027': '\u0027', // Straight Single Quotes
    '\u2018': '\u2019', // Comma Single Quotes
    '\u201B': '\u2019', // Curly Single Quotes Reversed
    '\u201A': '\u2019', // Low Curly Single Quotes
    '\u2039': '\u203A', // Angle Single Quotes
    '\u2035': '\u2032', // Prime Single Quotes
    '\u0060': '\u00B4'  // Prime Single Quotes
};

// Get all types of single quote.
const apostrophes = [].concat(
  Object.keys(singleQuotes),
  Object.values(singleQuotes)
)
  // Only unique quotes.
  .filter((v, i, s) => s.indexOf(v) === i)
  // In a string.
  .join('');

// Slightly adjust the current RegExps.
// [^\s\w]* match 0 or more of any char that is NOT a space or a alphanumeric
const afterWord = new RegExp('([a-z]s[' + apostrophes + '])[^\\s\\w]*$');
const apostrophe = new RegExp('(?:s[' + apostrophes + '])|(?:[' + apostrophes + ']s)$');
// Or: const apostrophe = new RegExp('s?[' + apostrophes + ']s?$');
// For extra lenience.
```

The first part to this is detecting and hence removing things that 'could' be quotation marks. Here is my approach (It's heavily based on your `quotation_step` approach):
```js
  // Inside `for(let i = 0; i < terms.length; i++)` but before the `is_possessive` function
    const term = terms.get(i);
    let text = term.text;

    // First detect open quotes before detecting apostrophes
    if (typeof singleQuotes[text[0]] !== 'undefined') {
      // Add the expected closing quotes to our inspection array.
      expectingClosers[expectingClosers.length] = new RegExp(
        // The quote, then not the quote, 0-9a-zA-Z_ or a space.
        singleQuotes[text[0]] + '[^' + singleQuotes[text[0]] + '\\s\\w]*$'
      );
      // Remove the found quote from the text (this isn't needed tbh.).
      text = text.slice(1);
    }

    // Pre checking for quotes. e.g: Carlos'.’. -> Carlos'.’
    // Strip everything that is not an apostrophe, a-z0-9_ from the text
    text = text.replace(new RegExp('[^' + apostrophes + '\\w]+$'), '');

    // If the string ends with an expected closer.
    const closer = expectingClosers
      .map(regex => regex.test(text))
      .findIndex(e => e === true);

    // If the closer is found.
    if (!!~closer) {
      // Remove the closer from the 
      text = text.replace(expectingClosers[closer], '');
      delete expectingClosers[closer];
    }

    // Post checking for quotes. e.g: Carlos'. -> Carlos'
    // Strip everything that is not an apostrophe, a-z0-9_ from the text
    text = text.replace(new RegExp('[^' + apostrophes + '\\w]+$'), '');
```
This then should leave us with only the possible isolated possessive to check. using `is_possessive`.

I made a few adjustments to `is_possessive`:
```js
const is_possessive = function(terms, text, index) {
  const thisWord = terms.get(index);
  const nextWord = terms.get(index + 1);

  if (
    !(
      // `blacklist` are always contractions, not possessive
      blacklist.includes(text) === true ||
      // If no apostrophe s or s apostrophe, return
      apostrophe.test(text) === false ||
      // Pronouns can't be possessive
      thisWord.tags.Pronoun === true
    ) && (
      // "spencers'" - this is always possessive - eg "flanders'"
      afterWord.test(text) === true ||
      //Last word is possessive  - "better than spencer's"
      nextWord === undefined ||
      thisWord.tags.ClauseEnd === true ||
      // Next word is 'house'
      nextWord.tags.Noun === true ||
      //rocket's red glare
      (
          nextWord.tags.Adjective &&
          terms.get(index + 2) &&
          terms.get(index + 2).tags.Noun
      )
    )
  ) {
    return true;
  }

  return false;
};
```
These include making all the `if statements` in to a single if (not 100% sure why though 😁), I added a check for comma separated clauses `thisWord.tags.ClauseEnd === true` and moving the `get`s to the top of the function.

Any thoughts/improvements to suggest?

**Comment by scagood** [comment]
I will note I've completely ignored the `.quotations()` for now, because, it should be a reasonably simple change when we agree on the tagging situation.

**Comment by spencermountain** [comment]
Love it.
really nice job. That's exactly what I was thinking about. very nice solution.

right, the possessives are being tagged in both the `possessive_step`, and the `./data/regex.js`. -- yeah, that's just me being sloppy. Nice find. Feel free to move that around, if it makes your logic easier.

also - 10 points for teaching me about the `~` operator in javascript. I had no idea about that.

PR!
i'm going on vacation (to japan) next friday. If you can get this in before that, we can do a release. 
This is great stuff.
cheers

**Comment by Errogant** [comment]
It is worth noting that the second tutorial contains an instance of quotations() still not working correctly.

Code:
```
{
  let text='Hi I’m Troy Maclure. You may remember me from from such instructional videos as “Mothballing Your Battleship” and “Dig Your Own Grave and Save”.'
  //parse the text..
  let doc=nlp(text)
  //grab a part of it (2 quotations)
  let matches= doc.quotations()
  //transform them to upper-case
  matches.toUpperCase()
  //return ad-hoc data
  return matches.map(ts=>ts.out())
}
```

Output:
```
Array(1) [
  0: " “MOTHBALLING YOUR BATTLESHIP”"
]
```

This outputs only the first quote, but I believe it should output both? I couldn't find a good source of documentation to make me believe otherwise, and a modification to the code shows that this is the only phrase tagged as a quote.

Code:
```
{
  let text='Hi I’m Troy Maclure. You may remember me from from such instructional videos as “Mothballing Your Battleship” and “Dig Your Own Grave and Save”.'
  //parse the text..
  let doc=nlp(text)
  //grab a part of it (2 quotations)
 return doc.match('#Quotation').out('text')
}
```

Output:

```
" “Mothballing Your Battleship”"
```

**Comment by spencermountain** [comment]
hey, yeah thanks @Errogant. 
I've made a new issue [here](https://github.com/spencermountain/compromise/issues/500)
